mmv1/templates/terraform/constants/network_endpoints.go.erb

```diff
@@ -0,0 +1,111 @@
+    e := endpoint.(map[string]interface{})
+    ipAddress := e["ip_address"].(string)
+    port := e["port"].(int)
+    instance, _ := e["instance"].(string)
+    return NetworkEndpointsNetworkEndpoint{
+        IPAddress: ipAddress,
+        Port: port,
+        Instance: instance,
+    }
+}
+
+func NetworkEndpointsNetworkEndpointConvertToAny(endpoint NetworkEndpointsNetworkEndpoint) interface{} {
+    m := make(map[string]interface{})
+    m["ip_address"] = endpoint.IPAddress
+    m["port"] = endpoint.Port
+    m["instance"] = endpoint.Instance
+    return m
+}
+
+// Continues to read network endpoints as long as there are unread pages remaining
+func networkEndpointsPaginatedRead(d *schema.ResourceData, config *transport_tpg.Config, userAgent, url, project, billingProject, pt string) ([]interface{}, error) {
+	var allEndpoints []interface{}
+	for len(pt) > 0 {
+		res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+			Config:    config,
+			Method:    "POST",
+			Project:   billingProject,
+			RawURL:    fmt.Sprintf("%s?pageToken=%s", url, pt),
+			UserAgent: userAgent,
+		})
+		if err != nil {
+			return nil, transport_tpg.HandleNotFoundError(err, d, fmt.Sprintf("ComputeNetworkEndpoints %q", d.Id()))
+		}
+		resEndpoints := res["items"].([]interface{})
+		allEndpoints = append(allEndpoints, resEndpoints...)
+		pt, _ = res["nextPageToken"].(string)
+	}
+	return allEndpoints, nil
+}
+
+// Mutates the parent NEG by attaching or detaching endpoints in chunks. `url` determines if endpoints are attached or detached.
+// The last page is not processed, but instead returned for the Create/Delete functions to write.
+func networkEndpointsPaginatedMutate(d *schema.ResourceData, endpoints []interface{}, config *transport_tpg.Config, userAgent, url, project, billingProject string, chunkSize int, returnLastPage bool) ([]interface{}, error) {
+	// Pull out what this mutation is doing - either attachNetworkEndpoints or detachNetworkEndpoints
+	verb := url[len(url)-len("attachNetworkEndpoints"):]
+    id, err := tpgresource.ReplaceVars(d, config, "{{project}}/{{zone}}/{{network_endpoint_group}}/endpoints")
+	if err != nil {
+		return nil, fmt.Errorf("Error constructing id: %s", err)
+	}
+    i := 0
+	for ; i < len(endpoints); i += chunkSize {
+		j := i + chunkSize
+		if j > len(endpoints) {
+			if returnLastPage {
+				break
+			}
+			j = len(endpoints)
+		}
+		body := map[string]interface{}{"networkEndpoints": endpoints[i:j]}
+		res, err := transport_tpg.SendRequest(transport_tpg.SendRequestOptions{
+			Config:    config,
+			Method:    "POST",
+			Project:   billingProject,
+			RawURL:    url,
+			UserAgent: userAgent,
+			Body:      body,
+			Timeout:   d.Timeout(schema.TimeoutDefault),
+		})
+		if err != nil {
+			return nil, fmt.Errorf("Error during %s: %s", verb, err)
+		}
+
+		err = ComputeOperationWaitTime(
+			config, res, project, verb, userAgent,
+			d.Timeout(schema.TimeoutDefault))
```

Comment:
This should probably use the appropriate Create / Delete timeout rather than the default timeout, I think?

Reply:
Ack, I used default initially since we don't set specific timeouts, but we can pick the right one based on `verb`.